pennylane_lightning/src/util/Memory.hpp

```diff
@@ -0,0 +1,188 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <limits>
+#include <memory>
+#include <new>
+
+#include "BitUtil.hpp"
+#include "TypeList.hpp"
+
+namespace Pennylane::Util {
+/**
+ * @brief Custom aligned allocate function. As appleclang does not support
+ * std::aligned_alloc in Mac OS 10.14, we use posix_memalign function.
+ *
+ * Note that alignment must be larger than max_align_t. Otherwise, the behavior
+ * is undefined.
+ *
+ * @param alignment Alignment value we want for the data pointer
+ * @param bytes Number of bytes to allocate
+ * @return Memory pointer
+ */
+inline auto alignedAlloc(uint32_t alignment, size_t bytes) -> void * {
+#if defined(__clang__)
+    /* Apple clang does not support std::aligned_alloc in Mac 10.14.
+     * Thus we use Posix function instead. */
+    void *p;
+    posix_memalign(&p, alignment, bytes);
+    return p;
+#elif defined(_MSC_VER)
+    return _aligned_malloc(bytes, alignment);
+#else
+    return std::aligned_alloc(alignment, bytes);
+#endif
+}
+
+/**
+ * @brief Free memory allocated by alignedAlloc.
+ *
+ * @param p Pointer to the memory location allocated by aligendAlloc
+ */
+inline void alignedFree(void *p) {
+#if defined(__clang__)
+    return ::free(p); // NOLINT(hicpp-no-malloc)
+#elif defined(_MSC_VER)
+    return _aligned_free(p);
+#else
+    return std::free(p);
+#endif
+}
+
+/**
+ * @brief C++ Allocator class for aligned memory.
+ *
+ * @tparam T Datatype to allocate
+ */
+template <class T> struct AlignedAllocator {
+    uint32_t alignment_;
+    using value_type = T;
+
+    /**
+     * @brief Constructor of AlignedAllocator class
+     *
+     * @param alignment Memory alignment we want.
+     */
+    constexpr explicit AlignedAllocator(uint32_t alignment)
+        : alignment_{alignment} {
+        // We do not check input now as it doesn't allow the constructor to be
+        // a constexpr.
+        // TODO: Using exception is allowed in GCC>=10
```

Comment:
Should we bump the GCC builder image? Ubuntu 22.04 should have GCC 11 by default.

Reply:
As I mention below, there will be a subtle issue in `cibuildwheel`.